### DISCUSSION TITLE
Flashing issue on module with 8MB PSRAM

### DISCUSSION THREAD

**Discussion by withoutgettingwet** [jomjol/AI-on-the-edge-device#1289] [General]
### The Problem

After flashing the firmware, nothing appears to happen. I found a similar issue, but the person reporting it could not even get output from the serial console, whereas I was able to (below)

This is occurring with two separately packaged ESP32-S boards, from the same seller. They appeared to work to some degree when first plugged in (the front light would turn on)

<details>
  <summary>Serial output</summary>
Rebooting...
ets Jul 29 2019 12:21:46

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:7568
load:0x40078000,len:15360
load:0x40080400,len:3744
entry 0x4008067c
I (27) boot: ESP-IDF 4.3.2 2nd stage bootloader
I (27) boot: compile time 17:20:48
I (27) boot: chip revision: 3
I (30) boot_comm: chip revision: 3, min. bootloader chip revision: 0
I (37) boot.esp32: SPI Speed      : 40MHz
I (42) boot.esp32: SPI Mode       : DIO
I (46) boot.esp32: SPI Flash Size : 4MB
I (51) boot: Enabling RNG early entropy source...
I (56) boot: Partition Table:
I (60) boot: ## Label            Usage          Type ST Offset   Length
I (67) boot:  0 nvs              WiFi data        01 02 00009000 00004000
I (74) boot:  1 otadata          OTA data         01 00 0000d000 00002000
I (82) boot:  2 phy_init         RF data          01 01 0000f000 00001000
I (89) boot:  3 ota_0            OTA app          00 10 00010000 001db000
I (97) boot:  4 ota_1            OTA app          00 11 001f0000 001db000
I (104) boot: End of partition table
I (109) boot_comm: chip revision: 3, min. application chip revision: 0
I (116) esp_image: segment 0: paddr=00010020 vaddr=3f400020 size=4df30h (319280) map
I (247) esp_image: segment 1: paddr=0005df58 vaddr=3ffb0000 size=020c0h (  8384) load
I (250) esp_image: segment 2: paddr=00060020 vaddr=400d0020 size=15e00ch (1433612) map
I (801) esp_image: segment 3: paddr=001be034 vaddr=3ffb20c0 size=029f8h ( 10744) load
I (806) esp_image: segment 4: paddr=001c0a34 vaddr=40080000 size=1c1f4h (115188) load
I (856) esp_image: segment 5: paddr=001dcc30 vaddr=50000000 size=00010h (    16) load
I (871) boot: Loaded app from partition at offset 0x10000
I (871) boot: Disabling RNG early entropy source...
I (883) psram: This chip is ESP32-D0WD
E (883) psram: PSRAM ID read error: 0xffffffff
E (883) cpu_start: Failed to init external RAM!
Re-enable cpu cache.

abort() was called at PC 0x40082dcf on core 0

Backtrace:0x40082c65:0x3ffe3ba0 0x4008f70d:0x3ffe3bc0 0x40096991:0x3ffe3be0 0x40082dcf:0x3ffe3c50 0x4007956d:0x3ffe3c80 |<-CORRUPTED

ELF file SHA256: 28356ed42df4c694
</details>

### Version

12.0.1

### Expected Behavior

The ESP module starts up and begins connecting to wifi

### Additional Context

Flashed using a purpose built 'mb' board from the seller.
Markings on the camera connector: TY OV2640-v1.0 
On the PSRAM: AP MEMORY / 6404L - 3SQR / 12205 / 150047G

**Comment by caco3** [comment]
seems to be a hardware issue:

```
E (883) psram: PSRAM ID read error: 0xffffffff
E (883) cpu_start: Failed to init external RAM!
Re-enable cpu cache.
```


either the hardware is not supported or defective (external RAM on module)!
You could try to flash another project and see if that works.

**Comment by stvogel** [comment]
I got exactly the same logs and problems on two boards of those: 
https://de.aliexpress.com/item/1005001978372728.html 
(also PSRAM 8MB)

**Comment by jomjol** [comment]
If it is written to be 8MB of PSRAM, it is not always inside.
There are several reports of users, that bought a 8MB board, but it turned out to be wrong and only 2 MB were inside!

**Comment by stvogel** [comment]
You are right ... checking for PSRAM these boards don't even have 2MB. When I check the PSRAM it reports a total of 0! WTF!

**Comment by withoutgettingwet** [comment]
The ones I have do have an 8mb psram chip on board, but I’m not sure whether or not it is really 8 mb or a mislabeled chip. When I get home I’ll check to see what it reads as

It does seem like this is a problem with the boards themselves and not the program after all.

I bought mine from AliExpress as well.

**Comment by caco3** [comment]
I also have several from Aliexpress, but they all work fine.
But AFAIK its impossible to know what you really order.

**Reply by stvogel** [reply]
Hi @caco3, could you send a link to the items you ordered.
As I just ordered three items of trash, I'd like to increase my chance to get some working EPS32-boards in this roulette-game :-/

**Reply by caco3** [reply]
The article seems no longer to be available (which is normal on Ali).
The same store has the same product now on https://de.aliexpress.com/item/1005004652465892.html?spm=a2g0o.store_pc_allProduct.8148356.2.12e84f6cLfzTEM&pdp_npi=2%40dis%21CHF%21CHF%204.95%21CHF%204.31%21%21%21%21%21%402103250d16678144344273617e15b4%2112000029988384077%21sh

The images look the same, but of course this does not need to mean anything!

**Comment by withoutgettingwet** [comment]
Since writing this, I've ordered another, from ebay this time. Unfortunately, it is also an 8MB psram that doesn't work (8mb according to the numbers engraved on the PSRAM chip - the listing said it was only 4) Same 6404L-3SQR psram, same error. If anyone manages to find a reliable seller, please tell me.

**Reply by caco3** [reply]
so is it 8MB instead of 4?
This would normally not be an issue (even better as you have double the RAM.
I see 2 scenarios that it does not work:
 - It is not really 8 but only 2 => can be checked by esptool or other firmware
 - The AI-on-the-edge-device firmware can not handle 8MB. It is unlikely and I would not think why this would be an issue

**Reply by withoutgettingwet** [reply]
>     * It is not really 8 but only 2 => can be checked by esptool or other firmware

Here's the output of esptool.py 
```Detecting chip type... ESP32
Chip is ESP32-D0WD-V3 (revision v3.0)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: 40:22:d8:f0:d5:a8
Stub is already running. No upload is necessary.
Manufacturer: d8
Device: 4016
Detected flash size: 4MB
Hard resetting via RTS pin... 
```
I'm not exactly sure what I'm looking at - I see the flash size is reported as 4MB but I don't see anything that is explicitly PSRAM, unless that's what it is referring to.

**Reply by caco3** [reply]
No, thats the flash sice, not the RAM (PSRAM).
It seems that the esptool is not able to show the PSRAM size, see https://github.com/espressif/esptool/issues/207 :(

Does the log of our firmware say something? I know it can detect 2MB and show an error.

You could try to write a little app by yourself, see https://thingpulse.com/esp32-how-to-use-psram/

In there it also states that 8MB is not usable as only 4MB can be addressed. Not sure if this is true for all ESPs/boards...

**Reply by withoutgettingwet** [reply]
I actually stumbled across the same post, and I've been trying to get that to work but I get nothing over serial once I've flashed it.

I've just tried using one of the example scripts in the ESP library for arduino and those don't seem to work either. as far as I can tell, it's just these boards.

**Reply by jomjol** [reply]
1) The esptool.py does not recognize PSRAM as this is an external chip, which is activated and controlled thru the firmware during the normal running of the software
2) Most of the ESP32CAM that are workin have an 8 MB PSRAM chip welded onto the board. Physically only 4 MB are accessible in an easy way and used by the AI-on-the-edge
3) If there is a problem with the PSRAM, than usually only a 2 MB chip is used (even if 4 or 8 is stated) or none at all. Beside a faulty chip as well.

**Reply by caco3** [reply]
> PSRAM .. accessible in an easy way

What does this mean?
Is it only addressable slowly?
Could we use it to eg. keep the processed images 8which we do not write to SD-Card to keep the write cycles low?

**Reply by jomjol** [reply]
Up to 4MB of RAM the esp32 can handle as if would be internal (malloc, ...). If you want to access more you need to use memory bank switching throug himem.
Example can be found here: https://github.com/espressif/esp-idf/tree/master/examples/system/himem

**Reply by caco3** [reply]
Thanks!

I quickly added the example to our project to test my devices. They all only have 4 MB PSRAM.

**Reply by withoutgettingwet** [reply]
So is it correct to say that this problem is due to something unusual/non-functional with these boards? 

RE: point 1: I've tried esptool.py, the espressif flasher, and the web browser based flasher (and both a board made to flash ESP32s and a USB to serial adapter).

I want to buy a standard 4MB one _somewhere_ but that's what I explicitly did with this new one, and it turned out to be an 8MB one anyway.

**Reply by caco3** [reply]
As @jomjol explained, it is ok to have  8MB. We just are only using 4 MB.

If the 8 MB one does not work, this IMO means:
 - The firmware can not handle this (That is very unlikely but I have no  8 MB board to confirm)
 - Your HW has a fault.
